client/state/plugins/premium/reducer.js

```diff
@@ -71,27 +70,20 @@ export function plugins( state = {}, action ) {
 			}
 			return state;
 		case SERIALIZE:
-			const processedState = {};
 			// Save the error state as a string message.
-			forEach( state, ( pluginList, key ) => {
-				processedState[ key ] = pluginList.map( item => {
+			return mapValues( state, pluginList =>
+				pluginList.map( item => {
 					if ( item.error !== null ) {
 						return Object.assign( {}, item, { error: item.error.toString() } );
 					}
 					return item;
-				} );
-			} );
-			return processedState;
-		case DESERIALIZE:
-			if ( ! isValidStateWithSchema( state, pluginInstructionSchema ) ) {
-				return {};
-			}
-			return state;
+				} )
+			);
 		default:
 			return state;
 	}
 }
-plugins.hasCustomPersistence = true;
+plugins.schema = pluginInstructionSchema;
```

Comment:
Just making a note, initially I was confused as to how this could behave the same. Here are the `withSchemaValidation` handlers:

https://github.com/Automattic/wp-calypso/blob/e6548c48daea3fcdd88548ce2bc3de5ff5fce81e/client/state/utils.js#L341-L352

It was surprising to me that `SERIALIZE` and `DESERIALIZE` are not both passed to the reducer (or neither). `SERIALIZE` is, but `DESERIALIZE` just returns the stored state.

Reply:
I totally agree @sirreal. Passing both SERIALIZE and DESERIALIZE to the wrapped reducer is one of the other patches that are queued up in my stash. (Right now, only SERIALIZE calls the wrapped reducer). Then we'll be able to layer the schema validation and custom serialization on top of each other:
```js
const reducer = withSchemaValidation( schema, ( state = new FooBar(), action ) => {
  switch ( action.type ) {
    case SERIALIZE:
      return state.toJSON(); // state is instance of `FooBar`
    case DESERIALIZE:
      // only valid state makes it all the way here: invalid state was caught by
      // `withSchemaValidation` and initialState was returned
      return new Foobar( state );
    default:
      ...
  }
} );
```
(function composition FTW!)

I'll add some testing instructions for the Happychat reducer.